### ISSUE TITLE
OpenMediaVault full container version ?

### ISSUE THREAD

**Issue by mcarbonneaux** [openmediavault/openmediavault#1437]
HomeAssistant has developped there ligth os to simplify and have more control on the global stability and simplicity = HAOS.
- https://github.com/home-assistant/operating-system
- https://www.home-assistant.io/installation/generic-x86-64

is based on buildroot os build system... https://buildroot.org/ 

is full container... all plugin are container (in that way you have more isolation from the host os)...
the base omv i think can be more integreted in that way...
and your are not dependant from base os upgrade.... and decision (switch from init to systemd...)...
you can be more up to date with kernel version...

you can separate hardware management (managed on the host, in form of api for example) from software management (in container)...

**Comment by ryecoaaron** [comment]
OMV requires real hardware and is really meant to be the host.  This would be a ground up rewrite for OMV.

**Comment by mcarbonneaux** [comment]
container can acces to the hardware, is juste an isolation... 
yes this the idea to of rewrite ... but i think the core omv6 are more consistant, and can be containerised...

**Comment by votdev** [comment]
Managing an own OS is exactly the opposite of that why OMV was started. In FreeNAS I had to invest more time in managing the FreeBSD ports than taking care of the features. I was annoyed by then and that is why OMV is based on Debian which is responsible for managing the distro things.

But this does not mean that OMV can not be used in container environments. Carving out various services into plugins will be the way to go to make that work. The modularisation is already on the todo list for OMV7.

**Comment by votdev** [comment]
After thinking about this I do not see an easy way to containerise OMV because all of the services it relies on. OMV is not stateless and making all configurations of all services persistent by storing them somehow on a shared volume seems to be overkill.

I think your origin intention is the service management of OMV. Maybe Cockpit it the better choice in this situation.

**Comment by votdev** [comment]
I don’t know buildroot enough, but I assume that you have to build every software package on your own. Choosing Debian as the core OS of OMV is to have thousands of existing packages and a package management out of the box.
I will not lower the great work of HASS, but you really can not compare that with OMV. HASS does not have to support file systems, audio, file services, replication, synchronisation, … In HASS you have to support Python and a plugin infrastructure and that is it.
I think that is like comparing apples with pears.

**Comment by mcarbonneaux** [comment]
bulidroot had this package list ready to use, and buildroot manage for you the dependency build for the buildroot package....
and debian carry tooo many depency and when you update the system with major update .... you have a major chance that broke many thing....

the haos is the infrastructure for the project... supervisor manage the containers (add-on or core product), the haos provide the needed base, mostly starting the os, managing container (docker)....
hass support many hardware (usb zigbee/zwave/ups....) that are needed to comunicate with iot equipement...

in omv the only part that manage storage hardware is the core part (md raid, fs, smart, lvm...) ... the rest are software that can be containerised... and use the fs... 
you can containerise smb, ftp, etc is juste softwara that manage file on fs... this container accessing the storage by mounting the part of the fs that e need in there container... 

hardware are completely accessible from container....

in the minimal os for omv you need only who manage docker, who manage hardrive, raid, and fs...
the rest can be manager by container...

cockpit is to generic... and too dependant of the alea of the os distribution...

**Comment by mcarbonneaux** [comment]
> shared volume seems to be overkill.
is not overkill

and you can use api (started on the host part) in place...

**Comment by ryecoaaron** [comment]
Just looking at the list of plugins and at least half won't work in a container.  I don't see a reason for a massive rewrite (don't have the man power) to support everything in a container when half of the plugins wouldn't run in a container.

**Comment by votdev** [comment]
I have to agree with @ryecoaaron that it does not make sense to containerise everything. I think using containerised software for plugins like MinIO and other already existing plugins make more sense.

**Comment by votdev** [comment]
Switching the base OS to something different than Debian is no option.

**Comment by mcarbonneaux** [comment]
my comment on each line :

> openmediavault-anacron :: yes but doesn't make sense for jobs that run on reboot
> openmediavault-apttool :: possible but pointless   `==> in container with host with buildroot no need`
> openmediavault-autoshutdown :: no `==> depend on what check you do`
> openmediavault-backup :: possible but needs to access os disk `==> this container can be pluged with mount on the base path of fs`
> openmediavault-borgbackup :: yes
> openmediavault-clamav :: yes but limited if not all filesystems are passed to container `==> you can pass the base path of the fs managed tree to the container...`
> openmediavault-compose :: no `==> yes you need to expose the docker socket in the container (eventualy to add somme CAP_), the supervisor of hass manage all the container of the host from a container...`
> openmediavault-cputemp :: no `==> yes, you can access to the /proc (virtualised version of the host version), and /sys, an dyou can access to device (haos for exemple as access to usb zigbee dongle from container)`
> openmediavault-diskclone :: no `==> it use dd on device... container can access to device...`
> openmediavault-diskstats :: no `==> yes, it's based on /proc/diskstats... juste need to exxpose the /proc`
> openmediavault-downloader :: yes
> openmediavault-fail2ban :: yes
> openmediavault-filebrowser :: yes (already is)
> openmediavault-flashmemory :: no `==> can be done in container but complex, this type of feature can be apified on the host part of the system`
> openmediavault-forkeddaapd :: yes
> openmediavault-hpraid :: no `==> can be containerised but i think this type of feature as to be in the host part (all hardware management in fact) as api...`
> openmediavault-kernel :: no `==> the kernel management as to be part of the core system (with api)...`
> openmediavault-kvm :: no `==> based on libvirt and libvirtd... can be containerised... the libvirtd also can be containerised (but i think it's must be part of the host os part... the libvirt client communicate with the libvirtd with unix socket... only need is to expose this socket to the container... or use tcp socket (it's posible with libvirtd)...`
> openmediavault-locate :: yes but limited if not all filesystems are passed to container `==> mount the fs base path to the container...`
> openmediavault-luksencryption :: no `==> must be on the host system and accesible by api`
> openmediavault-lvm2 :: no `==> the same`
> openmediavault-mergerfs :: no `==> the same`
> openmediavault-minidlna :: yes (version using docker exists)
> openmediavault-nut :: no `==> yes, only need to access to device, generaly usb device... haoss as this same plugin and is a conatiner...`
> openmediavault-omvextrasorg :: no `==> the base of the idea of containerisation of omv is to migrate the maximum of omvextra in container the avoid impact of each plugin in the omv ecosystem`
> openmediavault-onedrive :: yes (already is)
> openmediavault-owntone :: yes (already is)
> openmediavault-photoprism :: yes (already is)
> openmediavault-remotemount :: yes but difficult to use in other containers `==> i think must be part of the host system, and managed by api from core gui container`
> openmediavault-resetperms :: yes
> openmediavault-rsnapshot :: yes
> openmediavault-s3 :: yes (already is)
> openmediavault-sftp :: yes
> openmediavault-shairport :: yes
> openmediavault-sharerootfs :: no `==> all fs manipulation must be part of the host system as api...`
> openmediavault-snapraid :: no `==> can be in container is you expose the raid device... but can be part of the host system as api`
> openmediavault-snmp :: no `==> yes, beceause depend on /proc  and /sys, and eventualy some device... depend on what you whant to expose...`
> openmediavault-symlinks :: possible but pointless `==> i think in fact with the container system... can managed by mounting fs on specifique path on the container who need this fs... if not must be part of the host system by api`
> openmediavault-tftp :: yes
> openmediavault-tgt :: no `==> like other device management, no need to be a container, must be in host system managed by api`
> openmediavault-usbbackup :: no `==> can be a container start by the host on the event of the usb pluged ...`
> openmediavault-wakealarm :: no `==> it can be beceause it use /sys fs... i think can be done, but i'm not sure that need to be in container, i think more that y must be in the host system managed by api...`
> openmediavault-webdav :: yes but is currently using omv's nginx instance `==> you can use nginx instance dedicated on that... and you can have and ingress nginx that manage all you http traffic (omv gui, webdav, other you whant...  haoss manage there plugin by proxifying it through th ha gui and auth)`
> openmediavault-wetty :: yes (already is)
> openmediavault-wireguard :: yes
> openmediavault-wol :: no `==> why ? it's juste a packet sended by socket !`
> openmediavault-zfs :: no `==> like other fs/disk management, it must be in host system managed by api`

**Comment by mcarbonneaux** [comment]
> buildroot had this package list ready to use, and buildroot manage for you the dependency build for the buildroot package....

all the package maintenaid by the buildroot project:
- https://github.com/buildroot/buildroot/tree/master/package

for exemple php:
- https://github.com/buildroot/buildroot/tree/master/package/php

for python3: 
- https://github.com/buildroot/buildroot/tree/master/package/python3
plus most of the pip package are packaged with there respective dependency...

nginx: 
- https://github.com/buildroot/buildroot/tree/master/package/nginx

zfs:
- https://github.com/buildroot/buildroot/tree/master/package/zfs

lvm2:
- https://github.com/buildroot/buildroot/tree/master/package/lvm2

mdadm:
- https://github.com/buildroot/buildroot/tree/master/package/mdadm

**Comment by ryecoaaron** [comment]
You only look at what functionality is possible in docker and not how the plugin would interact with it. If I had to rewrite every plugin again to use a container and/or hit APIs just so it could live in a container, I will be moving on from this project.  I am a big fan of containers but containerizing just to containerize would do nothing for the project and steal more time from what there is a severe shortage of now.

**Comment by votdev** [comment]
I am on the same line with @ryecoaaron, it does not make sense to containerise things because it is fancy and hip at the moment. You have evaluate the pros and cons. And for the moment I do not see any sense for that except to use containers like it is done for some plugins because this really simplifies things.
And with this this discussion is closed.
